src/nupic/types/SdrProxy.hpp

```diff
@@ -91,17 +91,17 @@ class SDR_Proxy : public SDR
     ~SDR_Proxy() override
         { deconstruct(); }
 
-    SDR_dense_t& getDense() override {
+    SDR_dense_t& getDense() const override {
```

Comment:
now that SDR can be made const, SDRProxy lost one of its meanings. If we move the dimensions/resize functionality to SDR, we can prune Proxy. Thoughts?

Reply:
I'd rather not allow SDR's to change their dimensions, even if it stays the same size, since it could break assumptions of other code which watches the SDR.

Also: every time the SDR changes dimensions, it would need to invalidate the sparse data and reconvert.  With the current Proxy the sparse data is kept valid in both dimensions.

Reply:
> Also: every time the SDR changes dimensions, it would need to invalidate the sparse data  

I wouldn't worry about this one, sparse cache is just an sub-optimization, and by calling setDimensions() you expect reconversion.. 

> I'd rather not allow SDR's to change their dimensions,... since it could break assumptions of other code which watches the SDR.

Though this could be a problem, for code like 
```
const DIMX = sdr.getDimensions[0];
for (... < DIMX; ...)
```

so, we could disable the const version, and mention that behavior in the doc
```
size_t getDimensions() {...}
const size_t getDimensions() = delete;
```

Would that suffice? On the other hand, the merge of SDR and Proxy would be nice simplification & reduction.

Reply:
In Array, if the dimensions change I just create a new SDR.  But I got to thinking about this.  If the source region outputs an SDR  defined with one set of dimensions and the input buffer of the destination region is defined with a different set of dimensions, the source SDR is delivered to the destination and the original input SDR is replaced.  Is that going to be a problem?

The Link goes to a lot of trouble to make sure that both ends of a link match if the dimensions can be set at all.  But will the receiving region be expecting whatever the source sends it?

Reply:
> n Array, if the dimensions change I just create a new SDR 

For Region, do we need to support dimension changes at runtime? I'd say it's ok to be fixed once started

Reply:
If you want to  change the dimensions of an SDR before the program starts, call `sdr.initialize()`.  It's exposed for this reason.

> we can prune Proxy. Thoughts?

Sigh, I guess we can...  I'd like to keep it around for a little while though, maybe I can remember why I wrote it.  Also it's a good example of how to make subclasses of the SDR.